commerce_canadapost.module

```diff
@@ -22,26 +22,78 @@ const COMMERCE_CANADAPOST_LOGGER_CHANNEL = 'commerce_canadapost';
- *
- * Adds a field to the shipment entity for holding the tracking PIN number.
+ * Implements hook_form_FORM_ID_alter().
  */
-function commerce_canadapost_entity_base_field_info(EntityTypeInterface $type) {
-  if ($type->id() !== 'commerce_shipment') {
+function commerce_canadapost_form_commerce_shipment_canadapost_form_alter(&$form, FormStateInterface $form_state, $form_id) {
+  array_unshift($form['actions']['submit']['#submit'],'commerce_canadapost_commerce_shipment_form_submit');
+}
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ */
+function commerce_canadapost_form_commerce_checkout_flow_multistep_default_alter(&$form, FormStateInterface $form_state, $form_id) {
+  if(!isset($form['shipping_information']['shipments'])) {
+    return;
+  }
+
+  foreach ($form['shipping_information']['shipments'] as &$shipment) {
+    if (isset($shipment['field_actual_delivery'])) {
+      $shipment['field_actual_delivery']['#access'] = FALSE;
+    }
+    if (isset($shipment['field_attempted_delivery'])) {
+      $shipment['field_attempted_delivery']['#access'] = FALSE;
+    }
+    if (isset($shipment['field_expected_delivery'])) {
+      $shipment['field_expected_delivery']['#access'] = FALSE;
+    }
+    if (isset($shipment['field_mailed_on'])) {
+      $shipment['field_mailed_on']['#access'] = FALSE;
+    }
+  }
+}
+
+/**
+ * Functions.
+ */
+
+/**
+ * Provides a submit handler for the 'Save commerce shipment' button.
+ */
+function commerce_canadapost_commerce_shipment_form_submit($form, FormStateInterface $form_state) {
+  if (empty($form_state->getValue('tracking_code')[0]['value'])) {
     return;
   }
 
-  $fields = [];
+  /** @var \Drupal\commerce_shipping\Entity\Shipment $shipment */
+  $shipment = $form_state->getFormObject()->getEntity();
+  $current_tracking_pin = $shipment->get('tracking_code')->value;
+  $submitted_tracking_pin = $form_state->getValue('tracking_code')[0]['value'];
+  if ($submitted_tracking_pin === $current_tracking_pin) {
+    return;
+  }
+
+  /** @var \Drupal\commerce_canadapost\Api\TrackingService $tracking_service */
+  $tracking_service = \Drupal::service('commerce_canadapost.tracking_api');
+  $tracking_summary = $tracking_service->fetchTrackingSummary($submitted_tracking_pin);
+
+  $values = [];
+  if (!empty($tracking_summary['actual-delivery-date'])) {
+    $values['field_actual_delivery'][0]['value'] = new DrupalDateTime($tracking_summary['actual-delivery-date']);
+  }
+
+  if (!empty($tracking_summary['attempted-date'])) {
+    $values['field_attempted_delivery'][0]['value'] = new DrupalDateTime($tracking_summary['attempted-date']);
+  }
+
+  if (!empty($tracking_summary['expected-delivery-date'])) {
+    $values['field_expected_delivery'][0]['value'] = new DrupalDateTime($tracking_summary['expected-delivery-date']);
+  }
+
+  if (!empty($tracking_summary['mailed-on-date'])) {
+    $values['field_mailed_on'][0]['value'] = new DrupalDateTime($tracking_summary['mailed-on-date']);
+  }
```

Comment:
**NOT A CHANGE REQUEST**

*Readability*

I would also refactor this so that I wouldn't have to repeat the same code for each field:

```
$fields = [
  'actual-delivery-date' => 'field_actual_delivery',
  'attempted-date' => 'field_attempted_delivery',
  'expected-delivery-date' => 'field_expected_delivery',
  'mailed-on-date' => 'field-mailed_on',
];
foreach ($fields as $summary_field => $shipment_field) {
  if (!empty($tracking_summary[$summary_field])) {
    $values[$shipment_field] = new DrupalDateTime($tracking_summary[$summary_field]);
  }
}
```

Reply:
do you know if

```
if (!empty($tracking_summary[$summary_field])) {
    $values[$shipment_field] = new DrupalDateTime($tracking_summary[$summary_field]);
  }
```

will work?

i had a lot of trouble setting the values. hence the [0]['value']. i think the field might allow multiple values, so i couldnt figure it out

Reply:
Sorry, I meant to keep it as you had it `$values['field_expected_delivery'][0]['value']`.